File: src/sdks/javascript/src/metrics.ts
import type { ApmOptions, MetricAttributes, MetricRecord, KeyValue, AnyValue } from './types';

export class ApmMetrics {
  private options: Required<Pick<ApmOptions, 'endpoint' | 'applicationName'>> & ApmOptions;
  private queue: MetricRecord[] = [];
  private flushTimer: ReturnType<typeof setInterval> | null = null;

  constructor(options: ApmOptions) {
    this.options = {
      batchSize: 100,
      flushIntervalMs: 5000,
      maxRetries: 3,
      environment: 'development',
      ...options,
    };

    this.startFlushTimer();
  }

  counter(name: string, value: number, attributes?: MetricAttributes): void {
    this.record('counter', name, value, attributes);
  }

  gauge(name: string, value: number, attributes?: MetricAttributes): void {
    this.record('gauge', name, value, attributes);
  }

  histogram(name: string, value: number, attributes?: MetricAttributes): void {
    this.record('histogram', name, value, attributes);
  }

  private record(
    type: 'gauge' | 'counter' | 'histogram',
    name: string,
    value: number,
    attributes?: MetricAttributes
  ): void {
    const record: MetricRecord = {
      name,
      type,
      value,
      timestamp: Date.now() * 1_000_000,
      attributes: this.convertAttributes(attributes),
    };

    this.queue.push(record);

    if (this.queue.length >= (this.options.batchSize || 100)) {
      this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.queue.length === 0) return;

    const records = this.queue.splice(0, this.options.batchSize || 100);

    // Group by metric name
    const grouped = new Map<string, MetricRecord[]>();
    for (const record of records) {
      const existing = grouped.get(record.name) || [];
      existing.push(record);
      grouped.set(record.name, existing);
    }

    const metrics = Array.from(grouped.entries()).map(([name, recs]) => {
      const first = recs[0];
      const dataPoints = recs.map((r) => ({
        timeUnixNano: r.timestamp,
        asDouble: r.value,
        attributes: r.attributes,
      }));

      if (first.type === 'counter') {
        return {
          name,
          sum: {
            dataPoints,
            isMonotonic: true,
            aggregationTemporality: 2,
          },
        };
      }

      return {
        name,
        gauge: { dataPoints },
      };
    });

    const request = {
      resourceMetrics: [
        {
          resource: {
            attributes: [
              { key: 'service.name', value: { stringValue: this.options.applicationName } },
            ],
          },
          scopeMetrics: [
            {
              scope: { name: '@racelogic/apm-sdk' },
              metrics,
            },
          ],
        },
      ],
    };

    try {
      const headers: Record<string, string> = {
        'Content-Type': 'application/json',
      };

      if (this.options.apiKey) {
        headers['X-API-Key'] = this.options.apiKey;
      }

      if (this.options.applicationId) {
        headers['X-Application-Id'] = this.options.applicationId;
      }

      const response = await fetch(`${this.options.endpoint}/v1/metrics`, {
        method: 'POST',
        headers,
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        this.queue.unshift(...records);
      }
    } catch {
      this.queue.unshift(...records);
    }
  }

  private convertAttributes(attributes?: MetricAttributes): KeyValue[] | undefined {
    if (!attributes) return undefined;

    return Object.entries(attributes)
      .filter(([, v]) => v !== undefined)
      .map(([key, value]) => ({
        key,
        value: this.toAnyValue(value),
      }));
  }

  private toAnyValue(value: string | number | boolean | undefined): AnyValue {
    if (typeof value === 'string') return { stringValue: value };
    if (typeof value === 'number') {
      return Number.isInteger(value) ? { intValue: value } : { doubleValue: value };
    }
    if (typeof value === 'boolean') return { boolValue: value };
    return { stringValue: String(value) };
  }

  private startFlushTimer(): void {
    this.flushTimer = setInterval(() => {
      this.flush();
    }, this.options.flushIntervalMs || 5000);
  }

  async shutdown(): Promise<void> {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
    }
    await this.flush();
  }
}

File: src/sdks/python/src/racelogic_apm/metrics.py
"""APM Metrics implementation."""

import time
import threading
from typing import Any, Optional
from queue import Queue
from dataclasses import dataclass

import httpx

from .config import ApmConfig


@dataclass
class MetricRecord:
    name: str
    type: str  # "gauge", "counter", "histogram"
    value: float
    timestamp: int
    attributes: dict


class ApmMetrics:
    """Metrics collector for sending metrics to APM Collector."""

    def __init__(self, config: ApmConfig):
        self._config = config
        self._queue: Queue = Queue()
        self._client = httpx.Client(timeout=30.0)
        self._shutdown = False
        self._flush_thread = threading.Thread(target=self._flush_loop, daemon=True)
        self._flush_thread.start()

    def counter(self, name: str, value: int, **attributes: Any) -> None:
        """Record a counter metric (monotonically increasing value)."""
        self._record("counter", name, float(value), attributes)

    def gauge(self, name: str, value: float, **attributes: Any) -> None:
        """Record a gauge metric (point-in-time value)."""
        self._record("gauge", name, value, attributes)

    def histogram(self, name: str, value: float, **attributes: Any) -> None:
        """Record a histogram metric (distribution of values)."""
        self._record("histogram", name, value, attributes)

    def _record(
        self, metric_type: str, name: str, value: float, attributes: dict[str, Any]
    ) -> None:
        record = MetricRecord(
            name=name,
            type=metric_type,
            value=value,
            timestamp=int(time.time() * 1_000_000_000),
            attributes=attributes,
        )
        self._queue.put(record)

        if self._queue.qsize() >= self._config.batch_size:
            self._flush()

    def _convert_attributes(self, attributes: dict[str, Any]) -> list[dict]:
        result = []
        for key, value in attributes.items():
            if value is None:
                continue
            if isinstance(value, str):
                result.append({"key": key, "value": {"stringValue": value}})
            elif isinstance(value, bool):
                result.append({"key": key, "value": {"boolValue": value}})
            elif isinstance(value, int):
                result.append({"key": key, "value": {"intValue": value}})
            elif isinstance(value, float):
                result.append({"key": key, "value": {"doubleValue": value}})
            else:
                result.append({"key": key, "value": {"stringValue": str(value)}})
        return result

    def _flush(self) -> None:
        records: list[MetricRecord] = []
        while not self._queue.empty() and len(records) < self._config.batch_size:
            try:
                records.append(self._queue.get_nowait())
            except Exception:
                break

        if not records:
            return

        # Group by metric name
        grouped: dict[str, list[MetricRecord]] = {}
        for record in records:
            if record.name not in grouped:
                grouped[record.name] = []
            grouped[record.name].append(record)

        metrics = []
        for name, recs in grouped.items():
            first = recs[0]
            data_points = [
                {
                    "timeUnixNano": r.timestamp,
                    "asDouble": r.value,
                    "attributes": self._convert_attributes(r.attributes),
                }
                for r in recs
            ]

            if first.type == "counter":
                metrics.append(
                    {
                        "name": name,
                        "sum": {
                            "dataPoints": data_points,
                            "isMonotonic": True,
                            "aggregationTemporality": 2,
                        },
                    }
                )
            else:
                metrics.append({"name": name, "gauge": {"dataPoints": data_points}})

        request = {
            "resourceMetrics": [
                {
                    "resource": {
                        "attributes": [
                            {
                                "key": "service.name",
                                "value": {"stringValue": self._config.application_name},
                            }
                        ]
                    },
                    "scopeMetrics": [
                        {
                            "scope": {"name": "racelogic-apm"},
                            "metrics": metrics,
                        }
                    ],
                }
            ]
        }

        headers = {"Content-Type": "application/json"}
        if self._config.api_key:
            headers["X-API-Key"] = self._config.api_key
        if self._config.application_id:
            headers["X-Application-Id"] = self._config.application_id

        try:
            self._client.post(
                f"{self._config.endpoint}/v1/metrics",
                json=request,
                headers=headers,
            )
        except Exception:
            # Re-queue on failure
            for record in records:
                self._queue.put(record)

    def _flush_loop(self) -> None:
        while not self._shutdown:
            time.sleep(self._config.flush_interval_ms / 1000)
            if not self._shutdown:
                self._flush()

    def flush(self) -> None:
        """Flush all pending metrics."""
        self._flush()

    def shutdown(self) -> None:
        """Shutdown the metrics collector."""
        self._shutdown = True
        self._flush()
        self._client.close()
